src/state/kv_store.rs

```diff
@@ -0,0 +1,415 @@
+/*
+    Copyright © 2023, ParallelChain Lab
+    Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
+*/
+//! Defines the [KVStore] trait, which specifies the required interface for the key-value store provided 
+//! by the user.
+//! 
+//! Given a method to obtain the value for a given key in bytes, this module also provides methods to 
+//! obtain the variables stored in the key-value store, as described in [crate::state::block_tree].
+
+use std::fmt::Display;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use crate::hotstuff::types::QuorumCertificate;
+use crate::pacemaker::types::TimeoutCertificate;
+use crate::types::validators::{ValidatorSetUpdatesStatus, ValidatorSetUpdatesStatusBytes};
+use crate::types::{
+    basic::{AppStateUpdates, BlockHeight, ChildrenList, CryptoHash, Data, DataLen, Datum, ViewNumber}, 
+    block::Block,
+    validators::{ValidatorSet, ValidatorSetBytes, ValidatorSetState}
+};
+
+use super::paths;
+use super::utilities::combine;
+use super::write_batch::WriteBatch;
+
+pub trait KVStore: KVGet + Clone + Send + 'static {
+    type WriteBatch: WriteBatch;
+    type Snapshot<'a>: 'a + KVGet;
+
+    fn write(&mut self, wb: Self::WriteBatch);
+    fn clear(&mut self);
+    fn snapshot<'b>(&'b self) -> Self::Snapshot<'_>;
+}
+
+pub trait KVGet {
+    fn get(&self, key: &[u8]) -> Option<Vec<u8>>;
+
+    /* ↓↓↓ Block ↓↓↓  */
+
+    fn block(&self, block: &CryptoHash) -> Result<Option<Block>, KVGetError> {
```

Comment:
Is it really a good idea to define these functions in this trait? Shouldn't `KVGet` just represent a key-value store you can get from? Wouldn't it be better to define these functions in `BlockTree` and `BlockTreeSnapshot` instead?

Reply:
Do you mean that `KVGet` only has `fn get(&self, key: &[u8]) -> Option<Vec<u8>>`?

That's a good point, though re-designing it would be quite tricky. First, we'd need to duplicate methods in `BlockTree` and `BlockTreeSnapshot`. Secondly, `KVGetError` would lose its meaning, so we'd need to completely re-design the error types. 

This design is inherited from the previous versions of hotstuff-rs and to be honest I think it's ok. It enforces the intended key value store structure (there is no other acceptable structure, so this shouldn't be configurable), but leaves the database choice to the user.